sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -852,25 +855,148 @@ class Analyzer(
+              val newFilter = Filter(local.reduce(And), child)
+              fs += newFilter
+              newFilter
+            case xs =>
+              fs += child
+              child
+          }
+        case j: Join if j.joinType != Inner && fs.nonEmpty && j.find(fs.contains).isDefined =>
+          failAnalysis("accessing columns of outer query inside join is not supported")
+      }
+      (transformed, predicates.reduceOption(And))
+    }
+
+    /**
+     * Returns a resolved subquery and predicate that will be used to rewrite the IN subquery as
+     * semi join (predicate will be used as join condition).
+     */
+    private def rewriteInSubquery(
+        value: Expression,
+        subquery: LogicalPlan,
+        query: LogicalPlan): (LogicalPlan, Expression) = {
+      val (resolved, joinCondition) = extractCorrelatedPredicates(subquery, query)
+
+      // Extract the columns on the expression side.
+      val columns = value match {
+        case CreateStruct(cs) => cs
+        case c => Seq(c)
+      }
+
+      // The number of left and right expressions must be equal.
+      if (columns.length != resolved.output.length) {
+        throw new AnalysisException(s"the number of fields in value (${columns.length}) does" +
+          s" not match with the number of columns in subquery (${resolved.output.length})")
+      }
+      val conditions = joinCondition.toSeq ++ columns.zip(resolved.output).map {
+        case (e, a) =>
+          // Check the left and right dataTypes.
+          if (e.dataType != a.dataType) {
+            throw new AnalysisException(
+              s"data type of value (${e.dataType}) does not match with subquery (${a.dataType})")
+          }
+          EqualTo(e, a)
+      }
+
+      (resolved, conditions.reduceLeft(And))
+    }
+
+    def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
+      case f @ Filter(condition, child) if f.resolved =>
+        val (withSubquery, withoutSubquery) =
+          splitConjunctivePredicates(condition).partition(hasCorrelatedSubquery)
+
+        // Construct the pruned filter condition.
+        val newFilter: LogicalPlan = withoutSubquery match {
+          case Nil => child
+          case conditions => Filter(conditions.reduce(And), child)
+        }
+
+        // Filter the plan by applying left semi and left anti joins.
+        withSubquery.foldLeft(newFilter) {
+          case (p, Exists(sub)) =>
+            val (resolved, joinCondition) = extractCorrelatedPredicates(sub, p)
+            Join(p, resolved, LeftSemi, joinCondition)
+          case (p, Not(Exists(sub))) =>
+            val (resolved, joinCondition) = extractCorrelatedPredicates(sub, p)
+            Join(p, resolved, LeftAnti, joinCondition)
+          case (p, InSubQuery(value, sub)) =>
+            val (resolved, cond) = rewriteInSubquery(value, sub, p)
+            Join(p, resolved, LeftSemi, Some(cond))
+          case (p, Not(InSubQuery(value, sub))) =>
```

Comment:
For `A not IN (B, C)`, we need to consider the nullability of A and B. For example, if B or C is null, `A IN (B, null)` will be null, `A not in (B, null) ` should also be null. This requires a special join (it's usually called NullAwareAntiJoin).

We could not support that in this PR.